### DISCUSSION TITLE
Unit Testing and Testing Frameworks

### DISCUSSION THREAD

**Discussion by kcramakrishna** [Digital-Assistant/Digital_Assistant_Client#278] [Ideas]
Apparently there are 3 popular ways to have unit testing in javascript (and nodejs):

1. https://mochajs.org/ + https://www.chaijs.com/ + https://www.chaijs.com/plugins/sinon-chai/ (https://sinonjs.org/)
2. https://jestjs.io/
3. https://jasmine.github.io/

Apparently Jest is better suited for front end (and small projects):
https://dev.to/heroku/comparing-the-top-3-javascript-testing-frameworks-2cco
https://www.ponicode.com/blog/jest-versus-mocha-which-testing-framework-for-you
https://medium.com/@wheeler.katia/jest-vs-mocha-whats-the-difference-235df75ffdf3

There is also https://karma-runner.github.io/6.3/index.html but this seems more suited to Angular.

**General js testing**:
https://medium.com/information-and-technology/unit-testing-browser-extensions-bdd4e60a4f3d
https://www.codecademy.com/learn/learn-javascript-unit-testing
https://www.freecodecamp.org/news/how-to-start-unit-testing-javascript/
https://www.testim.io/blog/best-unit-testing-framework-for-javascript/

**Specifically testing Chrome/Browser Extensions**:
https://stackoverflow.com/questions/2869827/how-to-test-chrome-extensions
https://stackoverflow.com/questions/14798528/testing-browser-extensions/17370531#17370531
https://www.synopsys.com/blogs/software-security/creating-testable-modularized-chrome-extensions/
https://medium.com/information-and-technology/unit-testing-browser-extensions-bdd4e60a4f3d

There are a lot of articles comparing these. We need to evaluate all of these and any newer ones not listed above and standardise for our project. We may even end up using different tools for different parts of code. We may end up with Jest for front end and mocha+sinon+chai for backend.

**Test Coverage**:
https://www.merixstudio.com/blog/mocha-vs-jest/
Apparently jest has inbuilt test coverage.
Also we can look into test coverage tools like https://istanbul.js.org/ with mocha for test coverage.

@yureshwar : Please research them and document in this thread. You can probably try out all of them in the older code base i.e. write 5-10 unit tests in each of the frameworks and document the pros and cons. You can collaborate with @DharaniGongati on this.

Here are some articles to get you started:
https://www.educative.io/edpresso/mocha-vs-karma-vs-jasmine-what-to-use-when
https://stackshare.io/stackups/jasmine-vs-mocha-vs-qunit
https://www.codementor.io/@codementorteam/javascript-testing-framework-comparison-jasmine-vs-mocha-8s9k27za3
https://medium.com/dailyjs/javascript-test-runners-benchmark-3a78d4117b4
https://raygun.com/blog/mocha-vs-jasmine-chai-sinon-cucumber/
https://www.lambdatest.com/blog/jest-vs-mocha-vs-jasmine/
https://www.lambdatest.com/blog/top-javascript-automation-testing-framework/
https://2020.stateofjs.com/en-US/technologies/testing/ (This has some https://testing-library.com/ which seems to be popular - look into this more deeply)
https://github.com/orgs/testing-library/repositories
https://testing-library.com/docs/
https://knapsackpro.com/testing_frameworks/difference_between/jasmine/vs/mochajs
https://blog.qasource.com/software-development-and-qa-tips/what-are-the-differences-between-jasmine-mocha-and-cucumber-testing-framework

**Comment by yureshwar** [comment]
My observations on JEST

Writing a test unit is very easy in JEST and also for checking node applications for Unit tests it is very simple to run a command

Just write a test spec file like below

```
const sum = require('./sum');

test('adds 1 + 2 to equal 3', () => {
  expect(sum(1, 2)).toBe(3);
});
```

Run the following command to see the output
`yarn test or npm run test`

```
PASS  ./sum.test.js
✓ adds 1 + 2 to equal 3 (5ms)
```

It's pretty easy to maintain and you can easily track on the changes and do the graceful tests

**Reply by yureshwar** [reply]
Pros using jest:
1. Testing coverage report is available in Jest
2. Preconfigured features such as test runner, assertion library

Cons using jest:
1. Installing third-party libraries for backend specific testing because most of its tools are front-end specific

**Comment by yureshwar** [comment]
My views on Mocha

Mocha is also a good testing library but it has some cons in terms of performance and code needs to be improved. Writing test case on mocha is not straight forward like Jest. For example see as given below

```
var assert = require('assert');
describe('Array', function() {
  describe('#indexOf()', function() {
    it('should return -1 when the value is not present', function() {
      assert.equal([1, 2, 3].indexOf(4), -1);
    });
  });
});
```

**Comment by yureshwar** [comment]
My views on Testing Library([Link](https://testing-library.com/))

Testing library is also similar to Jest but it is not as popular as Jest and also for using it the docs are not so clear enough for a new developer to understand. You need to use specific libraries for each framework for example for react applications we need to use @testing-library/react library for writing the test cases. Hence for each and every framework there will be continuous learning for the developer to write the test cases in testing library

**Comment by kcramakrishna** [comment]
We have decided to use JEST for our project.
